### PR TITLE
Diagnostics

### PR DESCRIPTION
**Pull request by ardalis**
Fixes issue #62

### CODE REVIEW

**Comment by danroth27** [comment]
@glennc @Tratcher Please review

File: docs/fundamentals/diagnostics.rst [hunk 1]
```diff
@@ -1,6 +1,146 @@
...
+In ASP.NET 5, you configure the pipeline for each request in the ``Startup`` class's ``Configure()`` method (learn more about `configuration <configuration>`_). In order to add a simple error handling page, all that's required is to add a dependency on ``Microsoft.AspNet.Diagnostics`` to the project (and a ``using`` statement to ``Startup.cs``), and then add one line to ``Configure()`` in ``Startup.cs``:
+
+.. _diag-startup:
+
+.. literalinclude:: diagnostics/sample/src/DiagDemo/Startup.cs
+	:language: csharp
+	:linenos:
+	:emphasize-lines: 2,21
+
+The above code, which is built from the ASP.NET 5 Empty template, includes a simple mechanism for creating an exception on line 36. If a request includes a non-empty querystring parameter for the variable ``throw`` (e.g. a path of ``/?throw=true``), an exception will be thrown. Line 21 makes the call to ``UseErrorPage()`` with `ErrorPageOptions <https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics/ErrorPageOptions.cs>`_ set to ``ShowAll``. Using ``ErrorPageOption`` you can toggle the visibility of the following features of the error page:
```
Comment (danroth27) [inline docs/fundamentals/diagnostics.rst]: ErrorPageOption -> ErrorPageOptions

File: docs/fundamentals/diagnostics.rst [hunk 2]
```diff
@@ -1,6 +1,146 @@
...
+	- `The welcome page`_
+	
+`Browse or download samples on GitHub <https://github.com/aspnet/Docs/tree/master/docs/fundamentals/diagnostics/sample>`_.
+
+.. _configure-error-page
+
+Configuring an error handling page
+----------------------------------
+
+In ASP.NET 5, you configure the pipeline for each request in the ``Startup`` class's ``Configure()`` method (learn more about `configuration <configuration>`_). In order to add a simple error handling page, all that's required is to add a dependency on ``Microsoft.AspNet.Diagnostics`` to the project (and a ``using`` statement to ``Startup.cs``), and then add one line to ``Configure()`` in ``Startup.cs``:
```
Comment (danroth27) [inline docs/fundamentals/diagnostics.rst]: Add text clarifying that the error handling page is for development-time only.

File: docs/fundamentals/diagnostics.rst [hunk 3]
```diff
@@ -1,6 +1,146 @@
...
+Using the error page during development
+---------------------------------------
+
+The default error page will display some useful diagnostics information when an unhandled exception occurs within the web processing pipeline. The error page includes several tabs with information about the exception that was triggered and the request that was made. The first tab shows the stack trace:
+
+.. image:: diagnostics/_static/errorpage-stack.png
+
+The next tab shows the contents of the Querystring collection, if any:
+
+.. image:: diagnostics/_static/errorpage-query.png
```
Comment (danroth27) [inline docs/fundamentals/diagnostics.rst]: Crop the bottom of the image to remove unneeded whitespace.

File: docs/fundamentals/diagnostics.rst [hunk 4]
```diff
@@ -1,6 +1,146 @@
...
+
+In this case, you can see the value of the ``throw`` parameter that was passed to this request. This request didn't have any cookies, but if it did, they would appear on the Cookies tab. You can see the headers that were passed, here:
+
+.. image:: diagnostics/_static/errorpage-headers.png
```
Comment (danroth27) [inline docs/fundamentals/diagnostics.rst]: Crop the bottom of the image to remove unneeded whitespace.

**Comment by danroth27** [comment]
There are still AppInsights images in this PR that are not referenced and should be removed:
- azure-create-appinsight.png
- config-json.png
- nuget-pacakge-manager.png
- manage-nuget-packages.png

**Comment by ardalis** [comment]
The images are fixed in the AppInsights PR, just submitted.

File: docs/fundamentals/diagnostics.rst (hunk 4 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics.rst (hunk 3 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics.rst (hunk 2 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics.rst (hunk 1 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: fixed

**Comment by danroth27** [comment]
I'd rather not couple this PR with the AppInsights PR. Can we please remove the AppInsights images from this PR and add them in the correct PR?

**Comment by ardalis** [comment]
Images removed.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
+{
+    "webroot": "wwwroot",
+    "version": "1.0.0-*",
+
+    "dependencies": {
+        "Microsoft.AspNet.Server.IIS": "1.0.0-beta4",
+        "Microsoft.AspNet.Server.WebListener": "1.0.0-beta4",
+        "Microsoft.AspNet.Diagnostics": "1.0.0-beta4"
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json]: Add Kestrel. https://github.com/aspnet/WebSockets/blob/dev/test/AutobahnTestServer/Project.json#L8-L12

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                app.UseRuntimeInfoPage(); // default path is /runtimeinfo
+            }
+            else
+            {
+                // specify production behavior for error handling, for example:
+                // app.UseErrorHandler("/Home/Error");
+                // if nothing is set here, web server error page will be displayed
+            }
+
+            app.UseWelcomePage(new WelcomePageOptions() { Path = new PathString("/welcome") });
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: `app.UseWelcomePage("/welcome");`

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            if (string.Equals(env.EnvironmentName, "Development", StringComparison.OrdinalIgnoreCase))
+            {
+                app.UseErrorPage(ErrorPageOptions.ShowAll);
+
+                app.UseRuntimeInfoPage(); // default path is /runtimeinfo
+            }
+            else
+            {
+                // specify production behavior for error handling, for example:
+                // app.UseErrorHandler("/Home/Error");
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: Why is this commented out? You could send it to the welcome page for demo purposes.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                // if nothing is set here, web server error page will be displayed
+            }
+
+            app.UseWelcomePage(new WelcomePageOptions() { Path = new PathString("/welcome") });
+
+            app.Run(async (context) =>
+            {
+                if(context.Request.Query.ContainsKey("throw")) throw new Exception("Exception triggered!");
+                await context.Response.WriteAsync("Hello World!");
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: Can you send an HTML page with links to `/welcome' and '?throw'?
https://github.com/aspnet/Diagnostics/blob/dev/samples/ErrorHandlerSample/Startup.cs#L50-L53

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+  "profiles": {
+    "IIS Express": {
+      "commandName": "IISExpress",
+      "launchBrowser": true,
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json]: Also enable launchBrowser for `web` and `kestrel`

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Project_Readme.html [hunk 10]
```diff
@@ -0,0 +1,203 @@
+﻿<!DOCTYPE html>
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Project_Readme.html]: This is the default template readme, remove it.

File: docs/fundamentals/diagnostics/sample/global.json [hunk 11]
```diff
@@ -0,0 +1,6 @@
+﻿{
+    "projects": [ "src", "test" ],
+    "sdk": {
+        "version": "1.0.0-beta5-11584"
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/global.json]: 1.0.0-beta4

File: docs/fundamentals/diagnostics.rst [hunk 12]
```diff
@@ -1,6 +1,146 @@
...
+	errorPageOptions.ShowCookies = false;
+	app.UseErrorPage(errorPageOptions);
+
+Notice that the call to ``UseErrorPage()`` is wrapped inside an ``if`` condition that checks the current ``EnvironmentName``. This is a good practice, since you typically do not want to share detailed diagnostic information about your application publicly once it is in production. This check uses the ``ASPNET_ENV`` environment variable. If you are using Visual Studio 2015, you can customize the environment variables used when the application runs in the web application project's properties, under the Debug tab, as shown here:
+
+.. image:: diagnostics/_static/project-properties-env-vars.png
+	
+Setting the ``ASPNET_ENV`` variable to anything other than Development (e.g. Production) will cause the application not to call ``UseErrorPage()``, and thus any exceptions will be handled by the underlying web server (in this case, IIS Express) as shown here:
+
+.. image:: diagnostics/_static/oops-500.png
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics.rst]: This page comes from Microsoft.AspNet.Server.IIS, not IIS or IIS Express.

File: docs/fundamentals/diagnostics.rst [hunk 13]
```diff
@@ -1,6 +1,146 @@
...
+
+The next tab shows the contents of the Querystring collection, if any:
+
+.. image:: diagnostics/_static/errorpage-query.png
+
+In this case, you can see the value of the ``throw`` parameter that was passed to this request. This request didn't have any cookies, but if it did, they would appear on the Cookies tab. You can see the headers that were passed, here:
+
+.. image:: diagnostics/_static/errorpage-headers.png
+
+Finally, any environment variables defined for the server environment would be displayed on the Environment tab.
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics.rst]: Remove any reference to the environment tab, it's a leftover from Katana referring to the OWIN environment, not environment variables. https://github.com/aspnet/Diagnostics/issues/135

File: docs/fundamentals/diagnostics.rst [hunk 14]
```diff
@@ -1,6 +1,146 @@
...
+
+Once this is added to your ASP.NET application, you can browse to the specified path (``/runtimeinfo``) to see information about the runtime that is being used and the packages that are included in the application, as shown below:
+
+.. image:: diagnostics/_static/runtimeinfo-page.png
+
+The path for this page can be optionally specified in the call to ``UseRuntimeInfoPage()``. It accepts a `RuntimeInfoPageOptions <https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics/RuntimeInfoPageOptions.cs>`_ instance as a parameter, which has a ``Path`` property. For example, to specify a path of ``/info`` you would call ``UseRuntimeInfoPage()`` as shown here:
+
+.. code-block:: c#
+
+	app.UseRuntimeInfoPage(new RuntimeInfoPageOptions() { Path = new PathString("/info") });
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics.rst]: `app.UseRuntimeInfoPage("/info");`

File: docs/fundamentals/diagnostics.rst [hunk 15]
```diff
@@ -1,6 +1,146 @@
...
+
+Once included, this will handle all requests (by default) with a cool hello world page that uses embedded images and fonts to display a rich view, as shown here:
+
+.. image:: diagnostics/_static/welcome-page.png
+
+You can optionally configure the welcome page to only respond to certain paths. The code shown below will configure the page to only be displayed for the ``/welcome`` path (other paths will be ignored, and will fall through to other handlers):
+
+.. code-block:: c#
+
+	app.UseWelcomePage(new WelcomePageOptions() { Path = new PathString("/welcome") });
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics.rst]: app.UseWelcomePage("/welcome");

**Comment by Tratcher** [comment]
@danroth27 Is UseErrorHandler going to get it's own doc page?

**Comment by danroth27** [comment]
@Tratcher Yes, I think we should cover it as part of the Error Handling topic (#67).

**Comment by Tratcher** [comment]
Ok, then we should take it out of this sample, or at least put in a link to the other docs from that line in the sample.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json (hunk 5 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json]: done

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs (hunk 6 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: done - I'm sure I looked for an easier way when I wrote this... thanks for pointing it out.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs (hunk 7 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: It wouldn't make sense to show /welcome here as the error page, and for demo purposes this isn't running anything in a non-dev environment. We will have a separate article on UseErrorHandler and I added a TODO to link to it when it exists.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json [hunk 16]
```diff
@@ -0,0 +1,1146 @@
+{
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json]: @danroth27 do we want to check in lock files for samples? We don't do that elsewhere.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs (hunk 8 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Startup.cs]: done

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json (hunk 9 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json]: done

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Project_Readme.html (hunk 10 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Project_Readme.html]: done

File: docs/fundamentals/diagnostics/sample/global.json (hunk 11 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/global.json]: done

File: docs/fundamentals/diagnostics.rst (hunk 12 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics.rst (hunk 13 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json (hunk 16 above)
Comment (davidfowl) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json]: Agreed, we don't do it anywhere else.

File: docs/fundamentals/diagnostics.rst (hunk 14 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

File: docs/fundamentals/diagnostics.rst (hunk 15 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics.rst]: done

**Comment by ardalis** [comment]
Should be good to go now, @Tratcher thanks!

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json (hunk 16 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.lock.json]: Sorry, that should be in my .gitignore.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+        "Kestrel": "1.0.0-beta4"
+    },
+
+    "commands": {
+        "web": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.WebListener --server.urls http://localhost:5000",
+        "kestrel": "Microsoft.AspNet.Hosting --server Kestrel --server.urls http://localhost:12344"
+    },
+
+    "frameworks": {
+        "dnx451": { }/*,
```
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json]: `/*...*/` comments are no longer supported, remove them.  The sample should work on dnxcore50.

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json (hunk 9 above)
Comment (Tratcher) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/Properties/launchSettings.json]: launchBrowser is still missing for both kestrel and web, and there should also be a URL field.

**Comment by danroth27** [comment]
:shipit:

File: docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json (hunk 17 above)
Comment (ardalis) [inline docs/fundamentals/diagnostics/sample/src/DiagDemo/project.json]: Fixed

**Pull request merged**